implementations/rust/ockam/ockam_api/src/nodes/service/forwarder.rs

```diff
@@ -49,14 +94,135 @@ impl NodeManager {
+    Ok(Either::Right(r))
+}
+
+/// Configure the session for automatic recovery.
+fn enable_recovery(
+    session: &mut Session,
+    manager: Address,
+    ctx: Arc<Context>,
+    addr: MultiAddr,
+    alias: String,
+    auth: Arc<HashMap<MultiAddr, IdentityIdentifier>>,
+    mode: CredentialExchangeMode,
+) {
+    session.set_replacement(move |prev| {
+        let ctx = ctx.clone();
+        let addr = addr.clone();
+        let alias = alias.clone();
+        let auth = auth.clone();
+        let manager = manager.clone();
+        Box::pin(async move {
+            debug!(%addr, "creating new remote forwarder");
+            let f = async {
+                let a = replace_sec_chan(&ctx, &manager, prev, &addr, &auth, mode).await?;
+                RemoteForwarder::create_static(&ctx, Route::from(a.clone()), alias).await?;
+                Ok(a)
+            };
+            match timeout(MAX_RECOVERY_TIME, f).await {
+                Err(_) => {
+                    warn!(%addr, "timeout creating new remote forwarder");
+                    Err(ApiError::generic("timeout"))
+                }
+                Ok(Err(e)) => {
+                    warn!(%addr, err = %e, "error creating new remote forwarder");
+                    Err(e)
+                }
+                Ok(Ok(a)) => Ok(a),
+            }
+        })
+    })
+}
+
+async fn replace_sec_chan(
+    ctx: &Context,
+    manager: &Address,
+    prev: Address,
+    addr: &MultiAddr,
+    authorised: &HashMap<MultiAddr, IdentityIdentifier>,
+    mode: CredentialExchangeMode,
+) -> Result<Address> {
+    debug!(%addr, %prev, "recreating secure channel");
+    let req = DeleteSecureChannelRequest::new(&prev);
+    let req = Request::delete("/node/secure_channel").body(req).to_vec()?;
+    let vec: Vec<u8> = ctx.send_and_receive(manager.clone(), req).await?;
+    let mut d = Decoder::new(&vec);
+    let res: Response = d.decode()?;
+    if res.status() != Some(Status::Ok) && res.has_body() {
+        let e: Error = d.decode()?;
+        debug!(%addr, %prev, err = ?e.message(), "failed to delete secure channel");
+    }
+    let ids = authorised.get(addr).map(|i| vec![i.clone()]);
+    let mut req = CreateSecureChannelRequest::new(addr, ids, mode);
+    req.timeout = Some(MAX_CONNECT_TIME);
+    let req = Request::post("/node/secure_channel").body(req).to_vec()?;
+    let vec: Vec<u8> = ctx.send_and_receive(manager.clone(), req).await?;
+    let mut d = Decoder::new(&vec);
+    let res: Response = d.decode()?;
+    if res.status() != Some(Status::Ok) {
+        if res.has_body() {
+            let e: Error = d.decode()?;
+            warn!(%addr, %prev, err = ?e.message(), "failed to create secure channel");
+        }
+        return Err(ApiError::generic("error creating secure channel"));
+    }
+    let res: CreateSecureChannelResponse = d.decode()?;
+    let mad = res.addr()?;
+    if let Some(p) = mad.first() {
```

Comment:
Isn't in cleaner to convert this multiaddr to a route (we have a function for that) and check that there is only one hop?